### PR TITLE
AW-88 - seo first draft

### PR DESCRIPTION
**Pull request by meriouma**
Basically you have to add something like this in the ui binder
```
<seo:SeoWidget>
            <seo:Title>
                <ui:msg description="titleGaeView">GAE STUDIO - ArcBees</ui:msg>
            </seo:Title>
            <seo:Description>
                <ui:msg description="descGaeView">Description</ui:msg>
            </seo:Description>
            <seo:Keywords>
                <ui:msg description="keywordsGaeView">kw1,kw2,kw3</ui:msg>
            </seo:Keywords>
            <seo:Og>
                <seo:OgImage height="480" width="480">image url</seo:OgImage>
                <seo:OgType typeValue="WEBSITE"/>
            </seo:Og>
</seo:SeoWidget>
```

### CODE REVIEW

**Comment by jasonlemay** [comment]
YA MAN! LGTM

File: src/main/java/com/arcbees/seo/Seo.gwt.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module>
+	<inherits name='com.google.gwt.user.User'/>
```
Comment (Pacane) [inline src/main/java/com/arcbees/seo/Seo.gwt.xml]: wrong indent

**Comment by christiangoudreau** [comment]
Shouldn't we promote it to GWTP?

**Comment by christiangoudreau** [comment]
@Chris-V @spg

**Comment by christiangoudreau** [comment]
LGTM, but how inheritance would work?

File: src/main/java/com/arcbees/seo/Image.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.arcbees.seo.widget.OgImage;
+
+public class Image {
+    private final String url;
+    private final Integer width;
+    private final Integer height;
+    private final String mimeType;
+
+    public Image(String url, Integer height, Integer width, OgImage.MimeType type) {
```
Comment (Chris-V) [inline src/main/java/com/arcbees/seo/Image.java]: One param / line

File: src/main/java/com/arcbees/seo/widget/SeoWidget.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    private void inject(SeoElements seoElements) {
+        if (!isNullOrEmpty(seoElements.getTitle())) {
+            Document.get().setTitle(seoElements.getTitle());
+        }
+
+        setMetaTags(seoElements);
+    }
+
+    private void setMetaTags(SeoElements seoElements) {
```
Comment (Chris-V) [inline src/main/java/com/arcbees/seo/widget/SeoWidget.java]: What if on page A you have `og:image` configured but it's not configured on page B? When you go from page A to page B, the or:image metas won't be removed / updated. I think SeoWidget can keep a `List<MetaElement>` field and when the widget is detached, you detach the nodes contained in this list.

Comment (meriouma) [inline src/main/java/com/arcbees/seo/widget/SeoWidget.java]: Agree, was planning on adding that as an improvement if this proposal made sense!

Comment (Chris-V) [inline src/main/java/com/arcbees/seo/widget/SeoWidget.java]: gotcha!

**Comment by Chris-V** [comment]
LGTM

**Comment by meriouma** [comment]
The plan was to keep it out of GWTP I believe, and if needed create some bridge within GWTP. What do you mean by how inheritance would work?

**Comment by christiangoudreau** [comment]
Adding it to GWT-extensions then :D

**Pull request merged**